src/hooks/use_chat.rs

```diff
@@ -0,0 +1,202 @@
+    let session = use_session(cx);
+    let notification = use_notification(cx);
+    let room = use_room(cx);
+    let messages = use_messages(cx);
+    let threading_to = use_thread(cx);
+
+    let key_common_error_room_id = translate!(i18, "chat.common.error.room_id");
+
+    let messages_loading = use_ref::<bool>(cx, || false);
+    let limit_events_by_room = use_ref::<HashMap<String, u64>>(cx, || HashMap::new());
+    let from = use_ref::<Option<String>>(cx, || None);
+
+    let task_timeline = use_coroutine(cx, |mut rx: UnboundedReceiver<bool>| {
+        to_owned![
+            client,
+            room,
+            messages,
+            messages_loading,
+            limit_events_by_room,
+            from,
+            threading_to,
+            session,
+            notification
+        ];
+
+        async move {
+            while let Some(true) = rx.next().await {
+                messages_loading.set(true);
+
+                let current_room_id = room.get().id.clone();
+                let current_events = match limit_events_by_room.read().get(&current_room_id) {
+                    Some(c) => c,
+                    None => &(15 as u64),
+                }
+                .clone();
+
+                let session_data = match session.get() {
+                    Some(data) => data,
+                    None => {
+                        notification.set(NotificationItem {
+                            title: String::from(""),
+                            body: String::from(""),
+                            show: false,
+                            handle: NotificationHandle {
+                                value: NotificationType::None,
+                            },
+                        });
+
+                        return;
+                    }
+                };
+
+                let room_id = match RoomId::parse(&current_room_id) {
+                    Ok(id) => id,
+                    Err(_) => {
+                        notification.handle_error("{key_common_error_room_id}");
+                        return;
+                    }
+                };
+                let ms = messages.get().clone();
+
+                let (f, msg) = timeline(
+                    &client,
+                    &room_id,
+                    current_events,
+                    from.read().clone(),
+                    ms.to_vec(),
+                    session_data,
+                )
+                .await;
+
+                from.set(f);
+
+                info!("before write xxx");
+                messages.set(msg);
+                info!("after write xxx");
+                let mm = threading_to.get().clone();
+
+                if let Some(thread) = mm {
+                    let ms = messages.get().clone();
+                    let message = ms.iter().find(|m| {
+                        if let TimelineRelation::CustomThread(t) = m {
+                            if t.event_id.eq(&thread.event_id) {
```

Comment:
I'm not a big fan of doing some processing inside the find(it should just ...find), perhaps use the [`find_map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find_map) to return the `t` and do the processing outside(also less nesting ;))